src/utils/daos_hdlr.c

```diff
@@ -2467,12 +2457,20 @@ fs_copy(struct file_dfs *src_file_dfs,
 			D_GOTO(out, rc);
 		}
 	}
-
-	/* After going through all the entries, close the directory. */
-	if (file_closedir(src_file_dfs, src_dir))
-		fprintf(stderr, "Could not close '%s': %s\n",
-			dir_name, strerror(errno));
+err_file:
+	if (src_file_dfs->obj != NULL || src_file_dfs->fd != -1)
+		file_close(src_file_dfs, filename);
+	if (dst_file_dfs->obj != NULL || dst_file_dfs->fd != -1)
+		file_close(dst_file_dfs, filename);
 out:
+	/* don't try to closedir on something that is not a directory,
+	 * otherwise always close it before returning
+	 */
+	if (S_ISDIR(st_dir_name.st_mode)) {
+		if (file_closedir(src_file_dfs, src_dir))
+			fprintf(stderr, "Could not close '%s': %s\n",
+				dir_name, strerror(errno));
```

Comment:
Again, errno is not the right thing here.
```suggestion
                int err = file_closedir(src_file_dfs, src_dir);
                
                if (err != 0)
			fprintf(stderr, "Could not close '%s': %s\n",
				dir_name, strerror(err));
```

Reply:
@ashleypittman the rc is the errno, the file_closedir call returns an errno as its rc, so I changed errno to rc. dfs_closedir returns an errno as its rc, and the posix call sets errno, which is returned from file_closedir.

Reply:
file_closedir() does return the errno, but the log message is using errno again, which will be unset in the DAOS case.